src/server/pps/server/api_server.go

```diff
@@ -1477,185 +1502,157 @@ func (a *apiServer) jobManager(ctx context.Context, jobInfo *pps.JobInfo) {
 		df, err := newDatumFactory(ctx, pfsClient, jobInfo.Inputs, nil)
 		if err != nil {
 			return err
 		}
-		var inflightData int64
-		var processedData int64
-		var etcdProcessedData int64 // the value of processedData we've sent to etcd
+		tree := hashtree.NewHashTree()
+		var treeMu sync.Mutex
+
+		processedData := int64(0)
+		setProcessedData := int64(0)
 		totalData := int64(df.Len())
-		// This goro is responsible for updating job progress
-		// The values sent on this channel will be the number of datums
-		// processed so far.
-		progressCh := make(chan int64)
-		go func() {
-			var processed int64
-			for {
-				// We do the first update before receiving a progress,
-				// because we want to update the total number of datums.
+		var progressMu sync.Mutex
+		updateProgress := func(processed int64) {
+			progressMu.Lock()
+			defer progressMu.Unlock()
+			processedData += processed
+			// so as not to overwhelm etcd we update at most 100 times per job
+			if (float64(processedData-setProcessedData)/float64(totalData)) > .01 ||
+				processedData == 0 || processedData == totalData {
 				if _, err := col.NewSTM(ctx, a.etcdClient, func(stm col.STM) error {
 					jobs := a.jobs.ReadWrite(stm)
 					jobInfo := new(pps.JobInfo)
 					if err := jobs.Get(jobID, jobInfo); err != nil {
 						return err
 					}
-					// In case this goro races with the goro that sets
-					// this job as success, we don't want to overwrite
-					// the completed progress bar.
-					if jobInfo.DataProcessed != totalData {
-						jobInfo.DataProcessed = processed
-					}
-					jobInfo.DataTotal = int64(totalData)
+					jobInfo.DataProcessed = processedData
+					jobInfo.DataTotal = totalData
 					jobs.Put(jobInfo.Job.ID, jobInfo)
 					return nil
 				}); err != nil {
 					protolion.Errorf("error updating job progress: %+v", err)
-				}
-				select {
-				case processed = <-progressCh:
-				case <-ctx.Done():
-					// exit when the jobManager exits
-					return
+				} else {
+					setProcessedData = processedData
 				}
 			}
-		}()
+		}
+		// set the initial values
+		updateProgress(0)
 
-		tree := hashtree.NewHashTree()
-		files := df.Next()
-		for {
-			var resp hashtree.HashTree
-			var failed bool
-			if files != nil {
-				select {
-				case wp.DataCh() <- &datum{
-					files: files,
-				}:
-					files = df.Next()
-					inflightData++
-				case resp = <-wp.SuccessCh():
-					inflightData--
-				case <-jobFailedCh:
-					failed = true
-					inflightData--
-				}
-			} else {
-				if inflightData == 0 {
-					break
-				}
-				select {
-				case resp = <-wp.SuccessCh():
-				case <-jobFailedCh:
-					failed = true
-				}
-				inflightData--
-			}
-			if resp != nil {
-				if err := tree.Merge(resp); err != nil {
+		serviceAddr, err := a.workerServiceIP(ctx, deploymentName)
+		clientPool := sync.Pool{
+			New: func() interface{} {
+				conn, err := grpc.DialContext(ctx, fmt.Sprintf("%s:%d", serviceAddr, client.PPSWorkerPort), grpc.WithInsecure())
```

Comment:
Might wanna consider using the gRPC options that Matt was talking about this morning.  Otherwise might end up with unnecessary network errors.